Move FileNesting operations on projects to NodeBuilder's

rodrmoya opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/933930
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/918300
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/943240
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/923128
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/943248
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/948329

Commit:
[Ide] Move file nesting to ProjectFileNodeBuilder

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/918300

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -63,5 +65,15 @@ public static FilePath GetParentFile (Project project, FilePath inputFile)
 			return null;
 		}
+
+		public static bool HasChildren (Project project, FilePath inputFile)
```
These two will be called from code which is supposed to run fast and not be allocation heavy, as it happens when we build the solution tree. Can we drop the LINQ?

Also, can we not do O(N) lookup?

rodrmoya replied:
Yes, need to add a cache, will be doing that in the next commits

rodrmoya replied:
I've added now a sort of "cache", still in progress, but makes performance much better than before, where we were just making lots and lots of calls to the NestingRuleProvider's

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs:
```diff
@@ -134,7 +142,7 @@ public override int CompareObjects (ITreeNavigator thisNode, ITreeNavigator othe
 		{
 			ProjectFile file = (ProjectFile) dataObject;
-			return file.HasChildren;
+			return file.HasChildren ? true : FileNestingService.HasChildren (file.Project, file.FilePath);
```
```csharp
return file.HasChildren || FileNestingService.HasChildren (file.Project, file.FilePath);
```

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs:
```diff
@@ -143,6 +151,12 @@ public override void BuildChildNodes (ITreeBuilder treeBuilder, object dataObjec
 			if (file.HasChildren)
 				treeBuilder.AddChildren (file.DependentChildren);
+			else {
+				var children = FileNestingService.GetChildren (file.Project, file.FilePath);
```
GetChildren never returns null. Either we check for empty enumerable or change the API from IEnumerable.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -63,5 +65,15 @@ public static FilePath GetParentFile (Project project, FilePath inputFile)
+			return project.Files.Any (x => GetParentFile (project, x.FilePath) == inputFile);
+		}
+
+		public static IEnumerable<ProjectFile> GetChildren (Project project, FilePath inputFile)
```
Why return IEnumerable?

rodrmoya replied:
Because I wanted to use `ProjectFileCollection` internally, which only implements IEnumerable. Will look at improving it

Commit:
[Projects] Simplify FileNestingService API

Commit:
[Ide] Take into account file nesting rules in FolderNodeBuilder

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/943240

Commit:
[Ide] Add setting to toggle file nesting

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/923128

Therzok commented:
2019-07-08T18:10:35.7451160Z 1) MonoDevelop.Ide.PerfTests.TestSolutionLoad.TestLoad
2019-07-08T18:10:35.7451380Z Performance regression. Baseline: 8.248, Result: 14.326 (+0.74)

This is a pretty big regression (taking into account others have perf improvements)

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +54,32 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+
+		public static bool HasChildren (ProjectFile inputFile)
+		{
+			return inputFile.Project.Files.Any (x => GetParentFile (x) == inputFile);
```
O(N) lookup.

rodrmoya replied:
Yes, this is gone now

Commit:
[FileNesting] Apply file nesting based on project type

Commit:
[FileNesting] Keep file nesting information per project

Commit:
[Ide] Couple of fixes in the file nesting calls

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +55,137 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+			return projectFiles.TryGetValue (inputFile, out var nestingInfo) ? nestingInfo.Parent : null;
+		}
+
+		public ProjectFileCollection GetChidlrenForFile (ProjectFile inputFile)
```
Typo: GetChidlrenForFile -> GetChildrenForFile

slluis replied:
ProjectFileCollection allows adding/removing items. I don't think we support that in this data structure.

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreNestingRulesProvider.cs:
```diff
@@ -37,5 +40,11 @@ public AspNetCoreNestingRulesProvider ()
+		protected override bool AppliesToProject (Project project)
+		{
+			var dotnetProj = project as DotNetProject;
+			return dotnetProj != null && dotnetProj.MSBuildProject.GetReferencedSDKs ().FirstOrDefault (x => x.IndexOf ("Microsoft.NET.Sdk.Web", StringComparison.OrdinalIgnoreCase) != -1) != null;
```
[UsingMicrosoftNETSdkWeb](https://github.com/aspnet/websdk/blob/e8cb5ed1357d232251f26cf79413838ea62ba501/src/Web/Sdk/Sdk.props#L17) should be a more reliable way to do this.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -25,14 +25,17 @@
-	internal static class FileNestingService
+	public static class FileNestingService
 	{
+		static readonly ConcurrentDictionary<Project, ProjectNestingInfo> loadedProjects = new ConcurrentDictionary<Project, ProjectNestingInfo> ();
```
We indefinitely leak projects now. What we want here is a `ConditionalWeakTable`

rodrmoya replied:
yes, was going to add code to listen to ProjectClosed events, but I guess a `ConditionalWeakTable` looks even better, will change to that

rodrmoya replied:
Changed

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +55,137 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+
+		internal static ProjectFile InternalGetParentFile (ProjectFile inputFile)
 		{
+			if (inputFile.Project.UserProperties.HasValue ("FileNesting") && !inputFile.Project.UserProperties.GetValue<bool> ("FileNesting")) {
```
Any reason we check for the value existing first?

```suggestion
			if (!inputFile.Project.UserProperties.GetValue<bool> ("FileNesting", false)) {
```

rodrmoya replied:
Yeah, good one, fixed now, with a default of `true` instead, as that's why I was calling HasValue + GetValue, didn't see the overloaded method with the default

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +55,137 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+		}
 	}
+
+	class ProjectNestingInfo : IDisposable
```
```suggestion
	sealed class ProjectNestingInfo : IDisposable
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +55,137 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+
+	class ProjectNestingInfo : IDisposable
+	{
+		class ProjectFileNestingInfo
```
```suggestion
		sealed class ProjectFileNestingInfo
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -59,71 +59,74 @@ public NestingRule (NestingRuleKind kind, string appliesTo, IEnumerable<string>
 		public string AppliesTo { get; private set; }
 
-		bool CheckParentForFile (Project project, FilePath inputFile, FilePath parentFile)
+		ProjectFile CheckParentForFile (ProjectFile inputFile, FilePath parentFile)
```
From the looks of the callers, this is hot code. Avoid LINQ.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -59,71 +59,74 @@ public NestingRule (NestingRuleKind kind, string appliesTo, IEnumerable<string>
 			string parentFile, inputExtension;
 
-			inputExtension = Path.GetExtension (inputFile);
+			inputExtension = inputFile.FilePath.Extension;
```
We have [FilePath.HasExtension](http://source.monodevelop.com/#MonoDevelop.Core/MonoDevelop.Core/FilePath.cs,dff70377a8fc9935)

rodrmoya replied:
hmm, but this is to retrieve the extension, not to check if it has an extension, so what do you mean?

Therzok replied:
Old code used to work differently.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -59,71 +59,74 @@ public NestingRule (NestingRuleKind kind, string appliesTo, IEnumerable<string>
 		}
 
-		public FilePath GetParentFile (Project project, FilePath inputFile)
+		public ProjectFile GetParentFile (ProjectFile inputFile)
```
This is again, potentially unbounded hot code. Avoid string allocations where possible. ParentDirectory is done in a loop multiple times.

rodrmoya replied:
Fixed in https://github.com/mono/monodevelop/pull/8145/commits/dd64e9ba181098eebedb6949099b57f29fc440b0

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -121,8 +121,14 @@ static bool LoadFromFile (NestingRulesProvider provider)
 		}
 
-		public FilePath GetParentFile (Project project, FilePath inputFile)
+		protected virtual bool AppliesToProject (Project project) => true;
```
Shouldn't this use the Pad configuration property added [here](https://github.com/mono/monodevelop/pull/8145/files#diff-a07347c1eebf4a28b82422c42025f40fR52) ?

Therzok replied:
Either the FileNestingService has to use it (which seems wrong) or the node builders should. Marking as resolved and commenting there.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/FolderNodeBuilder.cs:
```diff
@@ -92,7 +93,7 @@ void GetFolderContent (Project project, string folder, out List<ProjectFile> fil
 				if (file.Subtype != Subtype.Directory) {
 					// If file depends on something other than a directory, continue
-					if (file.DependsOnFile != null && file.DependsOnFile.Subtype != Subtype.Directory)
+					if ((file.DependsOnFile != null && file.DependsOnFile.Subtype != Subtype.Directory) || FileNestingService.HasParent (file))
```
New pad option isn't being used?

rodrmoya replied:
Yes, it's used everywhere via `InternalGetParentFile`, which is what's used to fill the nesting information. These other methods just retrieve the already initialised data from the cache

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs:
```diff
@@ -96,11 +97,17 @@ public override object GetParentObject (object dataObject)
 					return groupUnder;
+			} else {
+				// File nesting
+				var parentFile = FileNestingService.GetParentFile (file);
```
New pad option isn't being used?

rodrmoya replied:
ditto

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs:
```diff
@@ -134,7 +141,7 @@ public override int CompareObjects (ITreeNavigator thisNode, ITreeNavigator othe
 		{
 			ProjectFile file = (ProjectFile) dataObject;
-			return file.HasChildren;
+			return file.HasChildren || FileNestingService.HasChildren (file);
```
New pad option isn't being used?

rodrmoya replied:
and ditto :)

Therzok commented:
Done with first review pass

Commit:
[FileNesting] Retrieve value with default in a single step

Commit:
[FileNesting] Seal ProjectNestingInfo class

Co-Authored-By: Marius Ungureanu <marius.ungureanu@xamarin.com>

Commit:
[FileNesting] Seal ProjectFileNestingInfo class

Co-Authored-By: Marius Ungureanu <marius.ungureanu@xamarin.com>

Commit:
[FileNesting] Fix typo

Commit:
[AspNetCore] Simplify check for determining application of ASP.NET nesting rules

Commit:
[FileNesting] Use a ConditionalWeakTable to store projects

Commit:
[FileNesting] Avoid LINQ in hot code

Commit:
[FileNesting] Retrieve file's ParentDirectory only once per file

Commit:
[FileNesting] Add code to react to file nesting being disabled/enabled

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -60,7 +60,8 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
 			if (!loadedProjects.TryGetValue (project, out var nestingInfo)) {
-				nestingInfo = loadedProjects [project] = new ProjectNestingInfo (project);
+				nestingInfo = new ProjectNestingInfo (project);
+				loadedProjects.Add (project, nestingInfo);
```
This can throw an exception.

Therzok replied:
Do a TryGetValue then a [GetOrCreateValue](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2.getvalue?view=netframework-4.8#System_Runtime_CompilerServices_ConditionalWeakTable_2_GetValue__0_System_Runtime_CompilerServices_ConditionalWeakTable__0__1__CreateValueCallback_) to ensure we're exception free from races.

rodrmoya replied:
Fixed with a single call to `GetValue`

Commit:
[FileNesting] Fix refreshes on file addition/removal

Commit:
[FileNesting] Update nesting information on file renames

Commit:
[FileNesting] Remove all children when removing files

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -29,6 +29,8 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Runtime.CompilerServices;
+using System.Web.UI.WebControls;
```
Huh?

rodrmoya replied:
Yeah, autocompleting to wrong types added this, fixed now

Commit:
[FileNesting] Remove unused namespaces

Commit:
[FileNesting] Convert PadOption to normal CommandItem

This allows us to store the setting in the root of the UserPrefs file, making
it easier to be retrieved from the FileNestingService in .Core.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +56,225 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+			Project.FileAddedToProject -= OnFileAddedToProject;
+			Project.FileRemovedFromProject -= OnFileRemovedFromProject;
+			Project.FileRenamedInProject += OnFileRenamedInProject;
+			Project.ParentSolution.UserProperties.Changed -= OnUserPropertiesChanged;
```
Maybe also clear `projectFiles`

Commit:
[FileNesting] Fix copy/paste issue

Commit:
[FileNesting] Fix case when opening solution with file nesting disabled

When so, we were not initializing the ProjectNestingInfo, which is where
we were connecting to the UserPropertiesChanged event, and failing to
do so meant we never got any notification and so users couldn't enable
file nesting.

Commit:
[FileNesting] Clear projectFiles dictionary on Dispose

Commit:
[FileNesting] Improve GetProjectNestingInfo method

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +56,210 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+				if (parent.Children == null) {
+					parent.Children = new ProjectFileCollection ();
+				}
+				if (!parent.Children.Contains (projectFile)) {
```
Maybe make this an O(1):

```csharp
var existing = parent.Children.GetFile (projectFile.FilePath);
if (existing == null) {
    parent.Children.Add (projectFile);
}
```

rodrmoya replied:
Coming in the next commit

Therzok reviewed (approved):
Other than comment, LGTM

Commit:
[FileNesting] Avoid using LINQ's Contains()

Commit:
[FileNesting] Add tests to ASP.NET Core addin

Now file nesting only works on specific project types, so the previous ones,
which were using a console application, were failing as no file nesting was
being applied.

Therzok commented:
cc @sevoku @mrward 👀

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +56,210 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+		public static bool IsEnabledForProject (Project project)
+		{
+			return project?.ParentSolution?.UserProperties?.GetValue<bool> ("MonoDevelop.Ide.FileNesting.Enabled", true) ?? true;
+		}
```
Still wondering if we should move this to MonoDevelop.Ide. Seems a bit odd to me to be looking for a MonoDevelop.Ide.FileNesting.Enabled property here.

Also presumably if it was in MonoDevelop.Ide the FileNestingService class would not need to be public?

rodrmoya replied:
Both done

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -52,16 +56,210 @@ static void HandleRulesProviderExtension (object sender, ExtensionNodeEventArgs
+		public static IEnumerable<ProjectFile> GetChildren (ProjectFile inputFile)
+		{
+			return GetProjectNestingInfo (inputFile.Project).GetChildrenForFile (inputFile);
+		}
```
Returning null from an IEnumerable seems wrong to me. GetChildrenForFile seems to return null when file nesting is disabled. Can it instead return an empty enumerable?

Commit:
[FileNesting] Move FileNestingService to .Ide

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,266 @@
+
+		public static IEnumerable<ProjectFile> GetChildren (ProjectFile inputFile)
+		{
+			return GetProjectNestingInfo (inputFile.Project).GetChildrenForFile (inputFile);
```
```suggestion
			return GetProjectNestingInfo (inputFile.Project).GetChildrenForFile (inputFile) ?? Array.Empty<ProjectFile> ();
```

mrward replied:
We could add this Array.Empty to the GetChildrenForFile method. Then no need for any null checks anywhere for the IEnumerable<ProjectFile> items.

Therzok replied:
GetChildrenForFile returns a ProjectFileCollection though.

mrward replied:
Good point 😄 I should have checked first.

rodrmoya replied:
I was going to revisit using a IEnumerable there, but don't really want to publicise the internal `ProjectFileCollection`, so I'm thinking about maybe using an `ImmutableList` internally and return a IReadOnlyList in the public API. I think it's better to check for null than calling `Any`, `Count`, etc on the `IEnumerable`, as this code gets called a lot.

Makes sense?

mrward commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,266 @@
+namespace MonoDevelop.Ide.Projects.FileNesting
+{
+	public static class FileNestingService
+	{
```
Internal?

rodrmoya replied:
Yes, moved to internal in the last commit

mrward commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs:
```diff
@@ -143,6 +148,12 @@ public override void BuildChildNodes (ITreeBuilder treeBuilder, object dataObjec
 				treeBuilder.AddChildren (file.DependentChildren);
+			else {
+				var children = FileNestingService.GetChildren (file);
+				if (children?.Any () ?? false) {
```
Would not be null with Marius's suggested code change.

Commit:
[FileNesting] Make FileNestingService internal

Commit:
[FileNesting] Don't use IEnumerable, surface internal collection now that service is internal

Commit:
[DotNetCore] Fix file nesting tests

Therzok commented:
LGTM

rodrmoya commented:
@monojenkins backport release-8.3

monojenkins commented:
@rodrmoya backporting to release-8.3 failed, the patch results in conflicts:

```
Applying: [Ide] Move file nesting to ProjectFileNodeBuilder
Applying: [Projects] Simplify FileNestingService API
Applying: [Ide] Take into account file nesting rules in FolderNodeBuilder
Applying: [Ide] Add setting to toggle file nesting
Applying: [FileNesting] Apply file nesting based on project type
Applying: [FileNesting] Keep file nesting information per project
Applying: [Ide] Couple of fixes in the file nesting calls
Applying: [FileNesting] Retrieve value with default in a single step
Applying: [FileNesting] Seal ProjectNestingInfo class
Applying: [FileNesting] Seal ProjectFileNestingInfo class
Applying: [FileNesting] Fix typo
Applying: [AspNetCore] Simplify check for determining application of ASP.NET nesting rules
Applying: [FileNesting] Use a ConditionalWeakTable to store projects
Applying: [FileNesting] Avoid LINQ in hot code
Applying: [FileNesting] Retrieve file's ParentDirectory only once per file
Applying: [FileNesting] Add code to react to file nesting being disabled/enabled
Applying: [FileNesting] Fix refreshes on file addition/removal
Applying: [FileNesting] Update nesting information on file renames
Applying: [FileNesting] Remove all children when removing files
Applying: [FileNesting] Remove unused namespaces
Applying: [FileNesting] Convert PadOption to normal CommandItem
Applying: [FileNesting] Fix copy/paste issue
Applying: [FileNesting] Fix case when opening solution with file nesting disabled
Applying: [FileNesting] Clear projectFiles dictionary on Dispose
Applying: [FileNesting] Improve GetProjectNestingInfo method
Applying: [FileNesting] Avoid using LINQ's Contains()
Applying: [FileNesting] Add tests to ASP.NET Core addin
Applying: [FileNesting] Move FileNestingService to .Ide
Applying: [FileNesting] Make FileNestingService internal
Using index info to reconstruct a base tree...
M	main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Tests/MonoDevelop.AspNetCore.Tests/LaunchSettingsTests.cs
Falling back to patching base and 3-way merge...
Auto-merging main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Tests/MonoDevelop.AspNetCore.Tests/LaunchSettingsTests.cs
CONFLICT (content): Merge conflict in main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Tests/MonoDevelop.AspNetCore.Tests/LaunchSettingsTests.cs
error: Failed to merge in the changes.
Patch failed at 0029 [FileNesting] Make FileNestingService internal
```

Please backport manually!

rodrmoya commented:
Backported manually at https://github.com/mono/monodevelop/pull/8495
